File: RewardRandomizer.Correlation/Correlator.cs
#nullable disable

using Microsoft.FSharp.Collections;
using RewardRandomizer;
using System;
using System.Collections.Generic;
using System.Linq;

namespace RewardRandomizer.Correlation
{
    public static class Correlator
    {
        public static IEnumerable<FSharpList<Reward>> ExtractSets(IEnumerable<Reward> list)
        {
            var bag = new HashSet<Reward>(list);

            var routes = new[]
            {
                (Route.Bartre, Route.Echidna),
                (Route.Ilia, Route.Sacae),
                (Route.Eirika, Route.Ephraim),
            };

            foreach (var (route1, route2) in routes)
            {
                // List all items in the A route
                foreach (var rA in bag.Where(x => x.route == route1).ToList())
                {
                    // Try to find a matching item in the B route (preferably one of the same target)
                    var rB = bag
                        .Where(x => x.route == route2)
                        .Where(x => x.item == rA.item)
                        .OrderBy(x => x.unit == rA.unit ? 1 : 2)
                        .ThenBy(x => Math.Abs(x.offset - rA.offset))
                        .DefaultIfEmpty(null)
                        .First();

                    if (rB != null)
                    {
                        // Matching item found - return A route and B route versions together, and remove them from the set
                        yield return ListModule.OfArray(new[] { rA, rB });
                        bag.Remove(rA);
                        bag.Remove(rB);
                    }
                }
            }

            foreach (var rA in bag.Where(x => x.route == Route.JoshuaAlive))
            {
                var rB = bag
                    .Where(x => x.route == Route.JoshuaDead)
                    .Where(x => x.item == rA.item)
                    .Single();
                var rC = bag
                    .Where(x => x.route == Route.Ephraim)
                    .Where(x => x.item == rA.item)
                    .Single();
                yield return ListModule.OfArray(new[] { rA, rB, rC });
                bag.Remove(rA);
                bag.Remove(rB);
                bag.Remove(rC);
            }

            // List all items in Eliwood Normal mode
            foreach (var rA in bag.Where(x => x.route == Route.EliwoodNormal).ToList())
            {
                // Try to find a matching item in each of the other routes
                Reward findClosest(Route route2) => bag
                    .Where(x => x.route == route2)
                    .Where(x => x.item == rA.item)
                    .OrderBy(x => Math.Abs(x.offset - rA.offset))
                    .DefaultIfEmpty(null)
                    .First();

                IEnumerable<Reward> getGroup()
                {
                    yield return rA;
                    if (findClosest(Route.HectorNormal) is Reward rB) yield return rB;
                    if (findClosest(Route.EliwoodHard) is Reward rC) yield return rC;
                    if (findClosest(Route.HectorHard) is Reward rD) yield return rD;
                }

                // Return all versions together, and remove them from the set
                var group = ListModule.OfSeq(getGroup());
                yield return group;
                foreach (var i in group)
                    bag.Remove(i);
            }

            // Find duplicates of items given to a specific unit (e.g. Percival's knight crest, depending on join chapter, or the angelic robe from The Liberation of Ilia, with dialogue depending on who survived)
            // These will be randomized as if they were a single item
            foreach (var g in bag.Where(x => x.route == Route.All && x.unit != 0).GroupBy(x => (x.unit, x.item)))
            {
                // Return items, grouped together
                yield return ListModule.OfSeq(g);
                // Remove each of them from the set
                foreach (var i in g)
                    bag.Remove(i);
            }

            foreach (var i in bag)
                if (i.route == Route.All)
                    yield return ListModule.Singleton(i);
                else if (i.route != Route.SpecialBehavior)
                    Console.Error.WriteLine("Unmatched: " + i);
        }
    }
}

File: RewardRandomizer/Correlator.fs
namespace RewardRandomizer

open System.Collections.Generic

module Correlator =
    let MutuallyExclusiveFilterSets = [
        let isOnRoute x r = r.Route = Some x
        let isNotOnRoute r = r.Route = None

        // Handle route splits
        for set in [
            [EarlyItem; LateItem]
            [Bartre; Echidna]
            [Ilia; Sacae]
            [EliwoodNormal; EliwoodHard; HectorNormal; HectorHard]
            [Tactician; NoTactician]
            [Lloyd; Linus]
            [Kenneth; Jerme]
            [Eirika; Ephraim]
        ] do
            yield ([for x in set do isOnRoute x], [])

        // Handle items that are neither
        yield ([isNotOnRoute], [])
    ]

    // Determines which items are the "same".
    // Groups of the "same" item are collected into sets.
    // Any items found are removed from the set.
    let GetMatches (rewards: HashSet<Reward>) required optional = [
        // Pick one of the routes/difficulty levels, and scan for items that appear in a special place there
        let primary_filter = List.head required
        let primary_rewards = rewards |> Seq.where primary_filter |> Seq.toList

        let remove item = ignore (rewards.Remove item)

        for p in primary_rewards do
            // This function will find the best match for this item on the route given.
            let find_matching filter =
                rewards
                |> Seq.where filter
                |> Seq.where (fun r -> r.ItemId = p.ItemId)
                |> Seq.sortBy (fun r -> [
                    if r.Unit = p.Unit then 1 else 2 // First, prefer an item who's "give to this unit" flag is the same as this one
                    abs (Set.minElement r.Offsets - Set.minElement p.Offsets) // In case of a tie, prefer the item whose location in the ROM is closer to this one
                ])
                |> Seq.tryHead

            // Find a match on each filter (both required and optional), and collect matches found into a list.
            let matching_set =
                (required @ optional)
                |> Seq.map find_matching
                |> Seq.choose id
                |> Set.ofSeq

            // Remove matched items from the pool for further operations.
            for x in matching_set do
                remove x

            // Make sure no required filters lacked a match.
            // If any of them did, this item won't be included in the randomizer.
            let missing_required_filters = seq {
                for filter in required do
                    let missing =
                        matching_set
                        |> Seq.where filter
                        |> Seq.isEmpty
                    if missing then yield filter
            }

            if Seq.isEmpty missing_required_filters then
                yield matching_set
    ]

    let ExtractAll (reward_source: seq<Reward>): Set<Reward> list = [
        // Create a HashSet to store objects to be processed
        let untagged_reward_set = HashSet reward_source
        // Get a list of filters that are mutually exclusive with each other (e.g. "is Sacae" and "is Ilia")
        // One of these will simply be "all routes, all difficulty levels"
        let reward_groups = MutuallyExclusiveFilterSets
        // Pass each to GetMatches, which will remove processed items from the HashSet,
        // and return the results
        for (required, optional) in reward_groups do
            yield! GetMatches untagged_reward_set required optional
    ]
